pilot/pkg/bootstrap/certcontroller.go

```diff
@@ -133,84 +129,172 @@ func (s *Server) initDNSCerts(hostname, customHost, namespace string) error {
+	caBundle := s.CA.GetCAKeyCertBundle().GetRootCertPem()
+	for {
+		select {
+		case <-stop:
+			return
+		case <-time.After(controller.NamespaceResyncPeriod):
+			newRootCert := s.CA.GetCAKeyCertBundle().GetRootCertPem()
+			if !bytes.Equal(caBundle, newRootCert) {
+				caBundle = newRootCert
+				certChain, keyPEM, err := s.CA.GenKeyCert(names, SelfSignedCACertTTL.Get(), false)
+				if err != nil {
+					log.Errorf("failed generating istiod key cert %v", err)
+				} else {
+					s.keyCertBundleWatcher.SetAndNotify(keyPEM, certChain, caBundle)
+				}
+			}
+		}
 	}
+}
 
-	// Save the certificates to ./var/run/secrets/istio-dns - this is needed since most of the code we currently
-	// use to start grpc and webhooks is based on files. This is a memory-mounted dir.
-	if err := os.MkdirAll(dnsCertDir, 0o700); err != nil {
-		return err
+// initCertificateWatches sets up watches for the dns certs.
+func (s *Server) initCertificateWatches(tlsOptions TLSOptions) error {
+	hasPluginCert := hasCustomTLSCerts(tlsOptions)
+	if !hasPluginCert && !s.EnableCA() && features.PilotCertProvider.Get() == IstiodCAProvider {
+		return nil
 	}
-	err = ioutil.WriteFile(dnsKeyFile, keyPEM, 0o600)
-	if err != nil {
-		return err
+	if hasPluginCert {
+		if err := s.keyCertBundleWatcher.SetFromFilesAndNotify(tlsOptions.KeyFile, tlsOptions.CertFile, tlsOptions.CaCertFile); err != nil {
+			return fmt.Errorf("set keyCertBundle failed: %v", err)
+		}
+		// TODO: Setup watcher for root and restart server if it changes.
+		for _, file := range []string{tlsOptions.CertFile, tlsOptions.KeyFile} {
+			log.Infof("adding watcher for certificate %s", file)
+			if err := s.fileWatcher.Add(file); err != nil {
+				return fmt.Errorf("could not watch %v: %v", file, err)
+			}
+		}
+		s.addStartFunc(func(stop <-chan struct{}) error {
+			go func() {
+				var keyCertTimerC <-chan time.Time
+				for {
+					select {
+					case <-keyCertTimerC:
+						keyCertTimerC = nil
+						if err := s.keyCertBundleWatcher.SetFromFilesAndNotify(tlsOptions.KeyFile, tlsOptions.CertFile, tlsOptions.CaCertFile); err != nil {
+							log.Errorf("Setting keyCertBundle failed: %v", err)
+						}
+					case <-s.fileWatcher.Events(tlsOptions.CertFile):
+						if keyCertTimerC == nil {
+							keyCertTimerC = time.After(watchDebounceDelay)
+						}
+					case <-s.fileWatcher.Events(tlsOptions.KeyFile):
+						if keyCertTimerC == nil {
+							keyCertTimerC = time.After(watchDebounceDelay)
+						}
+					case err := <-s.fileWatcher.Errors(tlsOptions.CertFile):
+						log.Errorf("error watching %v: %v", tlsOptions.CertFile, err)
+					case err := <-s.fileWatcher.Errors(tlsOptions.KeyFile):
+						log.Errorf("error watching %v: %v", tlsOptions.KeyFile, err)
+					case <-stop:
+						return
+					}
+				}
+			}()
+			return nil
+		})
+	}
+
+	neverStop := make(chan struct{})
+	watchCh := s.keyCertBundleWatcher.AddWatcher()
+	if err := s.loadIstiodCert(watchCh, neverStop); err != nil {
```

Comment:
This flow is confusing. There are two possible upstream goroutines that could write into the watcher channel - one that reads from the TLS file and one that periodically polls the Istiod CA. Could these be set up *AFTER* the watcher is added and the watcher cb is initialized? For instance, in initDNSCerts, there is a small chance that the CA bundle changes, but since the watcher is not setup, it won't reload the gRPC server with the new DNS cert

Reply:
Not sure i understand what you mean. 

this is the flow now:

1. in initDnsCerts, it already sets key/certs synchronously 
1. `initCertificateWatches` runs after `initCertificateWatches` and we load the istiod cert here.
 >one that periodically polls the Istiod CA

It is 60s period that we ccompare the root cert and re generate key/cert if needed. This time is long that istiod have enough time to bootstrap.

Reply:
I see what you mean. just a nit: Still just a little confusing to read. Think it would be better to add and setup the watcher before invoking the SetAndNotify routines.